compiler/optimizations/gpuTransforms.cpp

```diff
@@ -198,7 +202,45 @@ GpuizableLoop::GpuizableLoop(BlockStmt *blk) {
 
   this->loop_ = toCForLoop(blk);
   this->parentFn_ = toFnSymbol(blk->getFunction());
+  this->shouldErrorIfNotGpuizable_ = determineIfShouldErrorIfNotGpuizable();
   this->isEligible_ = evaluateLoop();
+
+  // Ideally we should error out earlier than this with a more specific
+  // error message but here's a final fallback error.
+  // There's one use case we want to exempt, which is failure to
+  // gpuize a nested loop. In this case if there was a failure to gpuize
+  // the outer loop we already would have errored.
+  if(this->shouldErrorIfNotGpuizable_ &&
+    !this->isEligible_ &&
+    !isAlreadyInGpuKernel())
+  {
+    USR_FATAL(blk, "Loop containing assertOnGpu() is not gpuizable");
+  }
+}
+
+bool GpuizableLoop::determineIfShouldErrorIfNotGpuizable() {
+  CForLoop *cfl = this->loop_;
+  INT_ASSERT(cfl);
+
+  // The gpuizable check will kick in if `assertOnGpu()` appears in the
+  // body of the loop absent of any control flow.
+  // It's necessary to do this instead of just checking the first
+  // statement as by the time we get to gpuTransforms code may have
+  // been added to the start of the loop (for example to
+  // assign to the loop iteration variable if we're iterating
+  // over values rather than indices)
+  for_alist(expr, cfl->body) {
```

Comment:
I can't say I'm familiar with the transformations that the production compiler applies to the function body, but this won't work in conditions even if they're `param`, right? code like
```Chapel
config param forceGpu: bool;
if forceGpu {
    assertOnGpu()
}
```

Reply:
Might a flag be useful for this kind of thing? (when processing a primitive call to `chpl_assert_on_gpu`, set a flag on the enclosing loop or function, if any? It's a bit more involved, of course.

Reply:
GpuTransforms runs pretty late in the compilation. So if forceGpu is false I think the  assert would be removed by dead code elimination.

If it's true I'm not sure if we have some other optimization that will kick in and remove the if or not.

I'll check (if not I'll probably open a new Github issue about making this more cognizant of control flow). You'd still get the runtime check either way though.

Reply:
I just tried the following and sure enough some prior pass simplifies the `if` statement away and by the time we get to `GpuTransforms` everything is straight-line code so it just works (you'll get the error that the loop is not GPUizable):

```chpl
use GPU;
config param forceGpu = true;

proc directlyRecursiveFunc() : int { return 1 + directlyRecursiveFunc(); }
on here.gpus[0] {    
  foreach i in 0..10 {
    if(forceGpu) { assertOnGpu(); }
    directlyRecursiveFunc();
  }
}
```

And as I expected the assert will be dead code eliminated if `forceGpu == false`

I'm sure you could come up with some example with control flow that wouldn't work but I'm not too worried about getting this to be robust enough to handle all possible situations: the expectation is that users will just make this the first statement in their loop, long term I hope we can remove a lot of the things that would cause something to fail to GPUize, and even if it fails to do the compile-time check it will still function as a runtime check.